nengo_loihi/inputs.py

```diff
@@ -0,0 +1,132 @@
+from __future__ import division
+
+import collections
+
+import numpy as np
+import nengo
+from nengo.exceptions import SimulationError
+from nengo.utils.compat import is_integer
+
+
+class CxSpikeInput(object):
+    def __init__(self, n, label=None):
+        self.n = n
+        self.label = label
+
+        self.spikes = {}  # map sim timestep index to list of spike inds
+        self.axons = []
+        self.probes = []
+
+    def add_axons(self, axons):
+        assert axons.group is None
+        axons.group = self
+        self.axons.append(axons)
+
+    def add_probe(self, probe):
+        if probe.target is None:
+            probe.target = self
+        assert probe.target is self
+        self.probes.append(probe)
+
+    def add_spikes(self, ti, spike_idxs):
+        assert is_integer(ti)
+        ti = int(ti)
+        assert ti > 0, "Spike times must be >= 1 (got %d)" % ti
+        assert ti not in self.spikes
+        self.spikes[ti] = spike_idxs
+
+    def clear_spikes(self):
+        self.spikes.clear()
+
+    def spike_times(self):
+        return sorted(self.spikes)
+
+    def spike_idxs(self, ti):
+        return self.spikes.get(ti, [])
+
+
+class PESModulatoryTarget(object):
+    def __init__(self, target):
+        self.target = target
+        self.errors = collections.OrderedDict()
+
+    def clear(self):
+        self.errors.clear()
+
+    def receive(self, t, x):
+        assert len(self.errors) == 0 or t >= next(reversed(self.errors))
+        if t in self.errors:
+            self.errors[t] += x
+        else:
+            self.errors[t] = np.array(x)
+
+    def collect_errors(self):
+        for t, x in self.errors.items():
+            yield (self.target, t, x)
+
+
+class HostSendNode(nengo.Node):
```

Comment:
This one isn't really an input, but it does seem like it should be grouped with these other objects.  Maybe `io_objects.py`?

Reply:
Whether it's input or output depends on whether you're thinking from the host or chip perspective, so I think it makes sense to call it something that doesn't force you into one perspective or the other. `io_objects.py` sounds good to me :+1: